extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/ReadOnlyDirectory.java

```diff
@@ -0,0 +1,35 @@
+package io.druid.lucene.segment.loading;
+
+import io.druid.lucene.LuceneDirectory;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.store.FSDirectory;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ */
+public class ReadOnlyDirectory implements LuceneDirectory {
+    private final IndexReader indexReader;
+
+    public ReadOnlyDirectory(File parentDir) throws IOException {
+        FSDirectory luceneDir = FSDirectory.open(parentDir.toPath());
+        indexReader = DirectoryReader.open(luceneDir);
+    }
+
+    @Override
+    public int numRows() {
+        return indexReader.numDocs();
+    }
+
+    @Override
+    public IndexReader getIndexReader() {
+        return indexReader;
```

Comment:
if caller calls IndexReader.close() on each thread, you will have failed other search threads with AlreadyClosedException.

The right pattern would be wrap IndexReader and on doClose(), call decRef instead. And call the actual close in the close() method. And for each getIndexReader() call, do incRef on the internal reader.

It is kinda tricky, let's have a test reproducing this scenario.